### PR TITLE
HDDS-2477. TableCache cleanup issue for OM non-HA.

### PR DESCRIPTION
**Pull request by bharatviswa504**
## What changes were proposed in this pull request?

In OM in non-HA case, the ratisTransactionLogIndex is generated by OmProtocolServersideTranslatorPB.java. And in OM non-HA validateAndUpdateCache is called from multipleHandler threads. So think of a case where one thread which has an index - 10 has added to doubleBuffer. (0-9 still have not added). DoubleBuffer flush thread flushes and call cleanup. (So, now cleanup will go and cleanup all cache entries with less than 10 epoch) This should not have cleanup those which might have put in to cache later and which are in process of flush to DB. This will cause inconsitency for few OM requests.

Example:

4 threads Committing 4 parts.

1st thread - part 1 - ratis Index - 3

2nd thread - part 2 - ratis index - 2

3rd thread - part3 - ratis index - 1

 

First thread got lock, and put in to doubleBuffer and cache with OmMultipartInfo (with part1). And cleanup is called to cleanup all entries in cache with less than 3. In the mean time 2nd thread and 1st thread put 2,3 parts in to OmMultipartInfo in to Cache and doubleBuffer. But first thread might cleanup those entries, as it is called with index 3 for cleanup.

 

Now when the 4th part upload came -> when it is commit Multipart upload when it gets multipartinfo it get Only part1 in OmMultipartInfo, as the OmMultipartInfo (with 1,2,3 is still in process of committing to DB). So now after 4th part upload is complete in DB and Cache we will have 1,4 parts only. We will miss part2,3 information.

 

So for non-HA case cleanup will be called with list of epochs that need to be cleanedup.

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-2477

## How was this patch tested?

Added UT.

### CODE REVIEW

File: hadoop-hdds/common/src/test/java/org/apache/hadoop/hdds/utils/db/cache/TestTableCacheImpl.java [hunk 1]
```diff
@@ -88,6 +89,55 @@ public void testPartialTableCache() {
   }
 
 
+  @Test
+  public void testPartialTableCacheWithNotContinousEntries() throws Exception {
+    int totalCount = 0;
+    int insertedCount = 3000;
+    for (long i=0; i<insertedCount; i+=2) {
```
Comment (anuengineer) [inline hadoop-hdds/common/src/test/java/org/apache/hadoop/hdds/utils/db/cache/TestTableCacheImpl.java]: what is the rationale of this +2 ?

File: hadoop-hdds/common/src/test/java/org/apache/hadoop/hdds/utils/db/cache/TestTableCacheImpl.java [hunk 2]
```diff
@@ -88,6 +89,55 @@ public void testPartialTableCache() {
...
+    // If cleanup policy is manual entries should have been removed.
+    if (cacheCleanupPolicy == TableCacheImpl.CacheCleanupPolicy.MANUAL) {
+      GenericTestUtils.waitFor(() -> {
+        return count - epochs.size() == tableCache.size();
+      }, 100, 10000);
+
+      // Check remaining entries exist or not and deleted entries does not
+      // exist.
+      for (long i = 0; i < insertedCount; i += 2) {
+        if (!epochs.contains(i)) {
```
Comment (anuengineer) [inline hadoop-hdds/common/src/test/java/org/apache/hadoop/hdds/utils/db/cache/TestTableCacheImpl.java]: same here, not sure why we are incrementing by +2.

File: hadoop-hdds/common/src/test/java/org/apache/hadoop/hdds/utils/db/cache/TestTableCacheImpl.java (hunk 1 above)
Comment (bharatviswa504) [reply hadoop-hdds/common/src/test/java/org/apache/hadoop/hdds/utils/db/cache/TestTableCacheImpl.java]: Just adding entries in the cache which are not in continuous order.

File: hadoop-hdds/common/src/test/java/org/apache/hadoop/hdds/utils/db/cache/TestTableCacheImpl.java (hunk 2 above)
Comment (bharatviswa504) [reply hadoop-hdds/common/src/test/java/org/apache/hadoop/hdds/utils/db/cache/TestTableCacheImpl.java]: Just adding entries in the cache which are not in continuous order.

File: hadoop-hdds/common/src/test/java/org/apache/hadoop/hdds/utils/db/cache/TestTableCacheImpl.java (hunk 1 above)
Comment (anuengineer) [reply hadoop-hdds/common/src/test/java/org/apache/hadoop/hdds/utils/db/cache/TestTableCacheImpl.java]: why?

Comment (bharatviswa504) [reply hadoop-hdds/common/src/test/java/org/apache/hadoop/hdds/utils/db/cache/TestTableCacheImpl.java]: And also my initial intention was to build epoch list of some count from this loop, and call cleanup cache. But later added a test with manual entries for epochs.

Comment (anuengineer) [reply hadoop-hdds/common/src/test/java/org/apache/hadoop/hdds/utils/db/cache/TestTableCacheImpl.java]: Are you trying to simulate multi-threaded action ? if so, should you be generating a random number and modding by the range? I am not sure +1 or +2 buys you any special testing here.

File: hadoop-hdds/common/src/test/java/org/apache/hadoop/hdds/utils/db/cache/TestTableCacheImpl.java (hunk 2 above)
Comment (anuengineer) [reply hadoop-hdds/common/src/test/java/org/apache/hadoop/hdds/utils/db/cache/TestTableCacheImpl.java]: I am not sure 0,2,4,6 satisfies the non-continuous order property you are seeking here.

File: hadoop-hdds/common/src/test/java/org/apache/hadoop/hdds/utils/db/cache/TestTableCacheImpl.java (hunk 1 above)
Comment (bharatviswa504) [reply hadoop-hdds/common/src/test/java/org/apache/hadoop/hdds/utils/db/cache/TestTableCacheImpl.java]: I just want to add entries in cache not in continous order and cleanup to be called with entries not in continous order. And as said my initial intention is to use this for loop to build epoch list for cleanup with some count. I can remove it if you prefer i++?

Comment (bharatviswa504) [reply hadoop-hdds/common/src/test/java/org/apache/hadoop/hdds/utils/db/cache/TestTableCacheImpl.java]: Updated test to use according to my original intention of i+=2

File: hadoop-hdds/common/src/test/java/org/apache/hadoop/hdds/utils/db/cache/TestTableCacheImpl.java (hunk 2 above)
Comment (bharatviswa504) [reply hadoop-hdds/common/src/test/java/org/apache/hadoop/hdds/utils/db/cache/TestTableCacheImpl.java]: Updated test to use according to my original intention of i+=2

File: hadoop-hdds/common/src/test/java/org/apache/hadoop/hdds/utils/db/cache/TestTableCacheImpl.java [hunk 3]
```diff
@@ -93,21 +93,21 @@ public void testPartialTableCache() {
   public void testPartialTableCacheWithNotContinousEntries() throws Exception {
     int totalCount = 0;
     int insertedCount = 3000;
+
+    int cleanupCount = 1000;
+
+    ArrayList<Long> epochs = new ArrayList();
     for (long i=0; i<insertedCount; i+=2) {
+      if (cleanupCount++ < 1000) {
```
Comment (anuengineer) [inline hadoop-hdds/common/src/test/java/org/apache/hadoop/hdds/utils/db/cache/TestTableCacheImpl.java]: The cleanupCount is set to 1000. Perhaps I am mistaken so I am reading this as -- cleanupCount is now 1001, and we check if it is less than 1000 ? I am still not sure why you wanted to check the non-continuous order, if it is to simulate threads, you probably need randomness?

Comment (bharatviswa504) [reply hadoop-hdds/common/src/test/java/org/apache/hadoop/hdds/utils/db/cache/TestTableCacheImpl.java]: Thanks for catching. I need to initialize cleanupCount to zero, as I want to have 1000 entries in epoch list. As I want to check behavior of cleanup with not continous entries with cache having entries like that.

Comment (anuengineer) [reply hadoop-hdds/common/src/test/java/org/apache/hadoop/hdds/utils/db/cache/TestTableCacheImpl.java]: Yes, but the source of the bug is that threads can be in inserting any entry under the current Epoch in different order. May I suggest that you create this list with a four loop and then do a simple shuffle. Then it does not matter if you are incrementing with 2 or 1. 

The point is that you need the cache entries out-of-order to verify your scenario.
Making it an even number list does not really qualify as an out-of-order list.

But it is fine; It is a minor detail; if you think this test cases addresses the issue sufficiently I am fine with going ahead with this change.

Comment (anuengineer) [reply hadoop-hdds/common/src/test/java/org/apache/hadoop/hdds/utils/db/cache/TestTableCacheImpl.java]: http://commons.apache.org/proper/commons-lang/javadocs/api-release/org/apache/commons/lang3/ArrayUtils.html#shuffle-boolean:A-

**Comment by bharatviswa504** [comment]
Tested out on 3 node cluster, with this patch, I was able to upload files with larger sizes, which is reported in HDDS-2356 MPU for larger files is failing with CompleteMultipartUpload error. I have tried upload of files up to 10GB, and able to upload successfully.


```
[root@bh-ozone-2 ozone-0.5.0-SNAPSHOT]# cp ./10g-1.file /mount/mpunov19/new10g.file
[root@bh-ozone-2 ozone-0.5.0-SNAPSHOT]# ls -al /mount/mpunov19/
total 30408708
drwxr-xr-x 2 root root        4096 Nov 20 14:49 .
drwxr-xr-x 4 root root          39 Nov 19 12:37 ..
-rw-r--r-- 1 root root 10737418240 Nov 19 13:40 10g-1.file
-rw-r--r-- 1 root root  1073741824 Nov 19 12:48 1g-1.file
-rw-r--r-- 1 root root  3221225472 Nov 19 13:29 3g-1.file
-rw-r--r-- 1 root root  5368709120 Nov 19 13:35 5g-1.file
-rw-r--r-- 1 root root 10737418240 Nov 20 14:49 new10g.file

Get the object again, and verified data is correct or not. 
[root@bh-ozone-2 ozone-0.5.0-SNAPSHOT]# ozones3api get-object --bucket mpunov19 --key new10g.file ./ozonenew10
{
    "AcceptRanges": "bytes", 
    "ContentType": "application/octet-stream", 
    "LastModified": "Wed, 20 Nov 2019 22:49:35 GMT", 
    "ContentLength": 10737418240, 
    "Expires": "Wed, 20 Nov 2019 22:53:44 GMT", 
    "CacheControl": "no-cache", 
    "Metadata": {}
}
[root@bh-ozone-2 ozone-0.5.0-SNAPSHOT]# diff 10g-1.file ozonenew10 
[root@bh-ozone-2 ozone-0.5.0-SNAPSHOT]
```

**Pull request merged**